_posts/2021-12-08-kafkasql-storage-and-security.md

```diff
@@ -0,0 +1,471 @@
+
+In this blog post, we will take a look at the KafkaSQL storage option in more detail. 
+We'll describe the bug and the solution, and provide a temporary workaround for users 
+until they are able to upgrade to the patched version `2.1.4.Final`.
+
+---
+
+# KafkaSQL Storage
+
+A major use-case for Apicurio Registry is to manage and distribute schemas for messages in Kafka topics. 
+This means that a lot of our users have a Kafka cluster available, so supporting it as a storage option has been a priority for us.
+
+KafkaSQL storage shares common logic with the standard SQL storage option, 
+but does not use a central SQL database to store Apicurio Registry data.
+Instead, each node uses a local SQL database. To keep the replicas in sync, each change in one of the
+local databases is replicated to the other nodes in the cluster using a Kafka topic.
+This makes the read operations on the data straightforward, and allows us to keep the commonality with the SQL persistence option.
+
+The design of the Kafka messages closely follows the tables that are used both in the local SQL database, 
+and the database for the standard SQL storage.
+
+Each Kafka message has a key and a corresponding value. Kafka itself is agnostic about their content, and treats them as a byte array.
+In the KafkaSQL storage, they are represented by objects that are serialized into a JSON String, 
+prepended with marker bytes used to recognize message type. 
+
+In general, each message value class has a corresponding key class. 
+Classes for the keys are mainly responsible for determining the partition that the message is going to reside in, 
+the tenant information in case of a multi-tenant deployment, and identifying information for the object that the message value represents. 
+For example `ArtifactVersionKey` contains group ID, artifact ID and a version identifiers for the given artifact version object.
+Choice of the partition key is important for maintaining consistency of the data between nodes, 
+because ordering of the messages is only guaranteed in the same partition.
+
+Classes fo the values contain a field with operation that the message represents, which includes `CREATE`, `UPDATE`, and `DELETE`.
+The other fields carry the rest of the information, such as the editable metadata for an artifact version.
+
+![KafkaSQL Storage Diagram](/images/guides/registry-storage-kafkasql-diagram.svg)
+*KafkaSQL storage option workflow diagram*
+
+When a change to a piece of data occurs, the originating nodes creates and sends the message to the Kafka topic,
+and blocks the user thread until the operation is processed. 
+Its local database is not modified directly. Each node only applies changes that are read form the Kafka topic.
+Therefore, user thread is unblocked only after the original message is read back from the topic by the node.
+This ensures that in case of a composite operation, messages are always processed in the correct order.
```

Comment:
Suggest adding `and critically in the *same* order for each node` to the end of the sentence.